rp2040-hal/src/i2c/peripheral.rs

```diff
@@ -0,0 +1,211 @@
+use core::{marker::PhantomData, ops::Deref};
+
+use crate::{
+    gpio::pin::bank0::BankPinId,
+    gpio::pin::{FunctionI2C, Pin, PinId},
+    resets::SubsystemReset,
+};
+use pac::{i2c0::RegisterBlock as I2CBlock, RESETS};
+
+use super::{Peripheral, SclPin, SdaPin, I2C};
+
+/// I2C bus events
+#[derive(Debug, PartialEq, Eq)]
+pub enum I2CEvent {
+    /// Start condition has been detected.
+    Start,
+    /// Restart condition has been detected.
+    Restart,
+    /// The controller requests data.
+    TransferRead,
+    /// The controller sends data.
+    TransferWrite,
+    /// Stop condition detected.
+    Stop,
+}
+
+impl<T, Sda, Scl> I2C<T, (Pin<Sda, FunctionI2C>, Pin<Scl, FunctionI2C>), Peripheral>
+where
+    T: SubsystemReset + Deref<Target = I2CBlock>,
+    Sda: PinId + BankPinId,
+    Scl: PinId + BankPinId,
+{
+    /// Configures the I2C peripheral to work in peripheral mode
+    ///
+    /// The bus *MUST* be idle when this method is called.
+    #[allow(clippy::type_complexity)]
+    pub fn new_peripheral(
+        i2c: T,
+        sda_pin: Pin<Sda, FunctionI2C>,
+        scl_pin: Pin<Scl, FunctionI2C>,
+        resets: &mut RESETS,
+        addr: u16,
+    ) -> I2CAsyncPeripheral<T, (Pin<Sda, FunctionI2C>, Pin<Scl, FunctionI2C>)>
+    where
+        Sda: SdaPin<T>,
+        Scl: SclPin<T>,
+    {
+        i2c.reset_bring_down(resets);
+        i2c.reset_bring_up(resets);
+
+        i2c.ic_enable.write(|w| w.enable().disabled());
+
+        // TODO: rp2040 supports 10bits addressing
+        //i2c_reserved_addr(addr)
+        i2c.ic_sar.write(|w| unsafe { w.ic_sar().bits(addr) });
+        // select peripheral mode & speed
+        i2c.ic_con.modify(|_, w| {
+            // run in fast mode
+            w.speed().fast();
+            // setup slave mode
+            w.master_mode().disabled();
+            w.ic_slave_disable().slave_enabled();
+            // hold scl when fifo's full
+            w.rx_fifo_full_hld_ctrl().enabled();
+            w.ic_restart_en().enabled();
+            w.tx_empty_ctrl().enabled()
+        });
+
+        // Clear FIFO threshold
+        i2c.ic_tx_tl.write(|w| unsafe { w.tx_tl().bits(0) });
+        i2c.ic_rx_tl.write(|w| unsafe { w.rx_tl().bits(0) });
+
+        // Enable DMA operations
+        //i2c.ic_dma_cr.write(|w| {
+        //    w.tdmae().enabled();
+        //    w.rdmae().enabled()
+        //});
+
+        // Enable IP
+        i2c.ic_enable.write(|w| w.enable().enabled());
+
+        I2CAsyncPeripheral {
+            i2c: Self {
+                i2c,
+                pins: (sda_pin, scl_pin),
+                mode: PhantomData,
+            },
+            state: State::Idle,
+        }
+    }
+}
+
+#[derive(Debug)]
```

Comment:
Could we move these `enum` and `struct` types up to the top of the file? I think it makes them easier to find.

Reply:
I usually try to keep the structure of the file as 
- impl-for-extern/parent-types
- repeated for each local type t:
  - type t
  - impl for type t

but I'm ok to move them at the top too.